### PR TITLE
Add subglacial runoff

### PR DESCRIPTION
**Pull request by irenavankova**
This adds subglacial runoff output from a file (e.g. MALI output) at the grounding line as freshwater volume flux at local freezing point. It follows closely what river runoff does.

### CODE REVIEW

File: components/mpas-ocean/src/analysis_members/Registry_conservation_check.xml [hunk 1]
```diff
@@ -197,6 +208,9 @@
 		<var name="accumulatedLandIceFrazilSalinityFlux" type="real" dimensions="Time" units="kg m^-2 s^-1"
 			 description="Salinity flux from frazil to Land Ice, given to coupler. Positive into the ocean."
 		/>
+	    <var name="accumulatedSubglacialRunoffSalinityFlux" type="real" dimensions="Time" units="kg m^-2 s^-1"
+			 description="Salinity flux from subglacial runoff. Positive into the ocean."
+		/>
```
Comment (xylar) [inline components/mpas-ocean/src/analysis_members/Registry_conservation_check.xml:213]: Here and below, make sure indentation is with tabs, not spaces. (Yes, I know it's stupid...)

File: components/mpas-ocean/src/Registry.xml [hunk 2]
```diff
@@ -691,6 +691,35 @@
 					description="The length scale of exponential decay of river runoff. Fluxes are multiplied by $e^{z/\gamma}$, where this coefficient is $\gamma$."
 					possible_values="Any positive real number."
 		/>
+		<nml_option name="config_flux_attenuation_coefficient_subglacial_runoff" type="real" default_value="0.001" units="m"
+					description="The length scale of exponential decay of subglacial runoff. Fluxes are multiplied by $e^{z/\gamma}$, where this coefficient is $\gamma$."
+					possible_values="Any positive real number."
+		/>
```
Comment (xylar) [inline components/mpas-ocean/src/Registry.xml:697]: It seems like we might need a logical config option that determines whether subglacial runoff is used at all.  It seems like that should control a package and variables should only be allocated and used if the package is on.

If we only were adding a few variables of size `nCells`, I wouldn't worry about this but `fractionAbsorbedSubglacialRunoff` also has `nVertLevels` so it's pretty big.

**Review by xylar — CHANGES_REQUESTED** [review]
@irenavankova, this looks like excellent work, as I always expect from you.

There are some small formatting fixes in the Registry to do.

The other issue, which we should discuss, is whether we want a namelist option and a package to prevent allocating subglacial runoff variables when they aren't used.

File: components/mpas-ocean/src/Registry.xml [hunk 3]
```diff
@@ -1023,6 +1052,10 @@
 					description="constant nondimensional salt transfer coefficient across the ice-ocean boundary layer"
 					possible_values="Any positive real number"
 		/>
+		<nml_option name="config_subglacial_runoff_mode" type="character" default_value="off"
+					description="Selects the mode in which subglacial runoff fluxes are implemented."
+					possible_values="'off', 'data'"
+		/>
```
Comment (xylar) [inline components/mpas-ocean/src/Registry.xml:1058]: I think this needs to get moved up to the same place as the other config options.  I would put it as the first config options before the others you added.

**Review by xylar — COMMENTED** [review]
One quick formatting suggestion.  I think this is otherwise ready to move to E3SM

**Comment by xylar** [comment]
Okay, no worries.  Ping me when the time is right to have another look.

**Comment by irenavankova** [comment]
@xylar, I made the subglacial stuff into a package now - I think it is ready for you to have a look again when you have time.

File: components/mpas-ocean/bld/build-namelist-section [hunk 4]
```diff
@@ -305,6 +312,7 @@ add_default($nl, 'config_land_ice_flux_ISOMIP_gammaT');
 add_default($nl, 'config_land_ice_flux_rms_tidal_velocity');
 add_default($nl, 'config_land_ice_flux_jenkins_heat_transfer_coefficient');
 add_default($nl, 'config_land_ice_flux_jenkins_salt_transfer_coefficient');
+add_default($nl, 'config_subglacial_runoff_mode');
```
Comment (xylar) [inline components/mpas-ocean/bld/build-namelist-section:315]: This option got moved to the `forcing` namelist group in the registry so it should be moved up here as well.

File: components/mpas-ocean/bld/namelist_files/namelist_definition_mpaso.xml [hunk 5]
```diff
@@ -1614,6 +1669,14 @@ Valid values: Any positive real number
...
 </entry>
 
+<entry id="config_subglacial_runoff_mode" type="char*1024"
+	category="land_ice_fluxes" group="land_ice_fluxes">
+Selects the mode in which subglacial runoff fluxes are computed.
+
+Valid values: 'off', 'data'
+Default: Defined in namelist_defaults.xml
+</entry>
+
```
Comment (xylar) [inline components/mpas-ocean/bld/namelist_files/namelist_definition_mpaso.xml:1679]: This should also be moved up to the `forcing` group.

File: components/mpas-ocean/src/Registry.xml [hunk 6]
```diff
@@ -3527,7 +3570,10 @@
 			 description="Divergence of transmission through interfaces of surface fluxes below the surface layer at cell centers. These are applied only to river runoff."
 			 packages="forwardMode;analysisMode"
 		/>
-
+		<var name="fractionAbsorbedSubglacialRunoff" type="real" dimensions="nVertLevels nCells Time" units="fractional"
+			 description="Divergence of transmission through interfaces of surface fluxes below the surface layer at cell centers. These are applied only to subglacial runoff."
+			 packages="dataSubglacialRunoffFluxPKG"
+		/>
```
Comment (xylar) [inline components/mpas-ocean/src/Registry.xml:3576]:
```suggestion
		/>

```
Maybe add back the blank line that was here before, just for visual separation between fields for different purposes?

File: components/mpas-ocean/src/shared/mpas_ocn_forcing.F [hunk 7]
```diff
@@ -180,6 +182,48 @@ subroutine ocn_forcing_build_fraction_absorbed_array(meshPool, statePool, forcin
            end do
         end do
 
+!  now do subglacial runoff separately
+        if ( trim(config_subglacial_runoff_mode) == 'data' ) then
+            if ( trim(config_sgr_flux_vertical_location) == 'top' ) then
```
Comment (xylar) [inline components/mpas-ocean/src/shared/mpas_ocn_forcing.F:187]: Would it make sense to add and `else` this this and raise an error if `config_sgr_flux_vertical_location` has an unexpected value?  Otherwise, I worry that a typo could run fine but have weird and hard-to-explain behavior.

File: components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F [hunk 8]
```diff
@@ -3411,6 +3439,37 @@ subroutine ocn_compute_KPP_input_fields(statePool, forcingPool, &
                       - exp( max(-100.0_RKIND, &
                                  -layerThickness(kmin, iCell)/ &
                            config_flux_attenuation_coefficient_runoff))
+         if (trim(config_subglacial_runoff_mode) == 'data') then
+             if (config_use_sgr_opt_kpp == 1) then
+                if ( trim(config_sgr_flux_vertical_location) == 'top' ) then
```
Comment (xylar) [inline components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:3444]: Similar to diagnostics, would it make sense to add and `else` this this and raise an error if `config_sgr_flux_vertical_location` has an unexpected value?  Otherwise, I worry that a typo could run fine but have weird and hard-to-explain behavior.

File: components/mpas-ocean/src/tracer_groups/Registry_ecosys.xml [hunk 9]
```diff
@@ -410,7 +410,7 @@
 				<var name="NH4SurfaceFluxRunoff" array_group="ecosysSurfaceFluxRunoffGRP" units="mmol N m^{-3} m s^{-1}"
 					description="Dissolved Ammonia Surface Flux Due to Runoff"
 				/>
-				<var name="FeSurfaceFluxRunoff" array_group="ecosysSurfaceFluxRunoffSurfaceFluxRunoffGRP" units="mmol Fe m^{-3} m s^{-1}"
+				<var name="FeSurfaceFluxRunoff" array_group="ecosysSurfaceFluxRunoffGRP" units="mmol Fe m^{-3} m s^{-1}"
```
Comment (xylar) [inline components/mpas-ocean/src/tracer_groups/Registry_ecosys.xml:413]: I'm sure this is right but probably should be in its own PR.  Do you want to check with Mat Maltrud about that?

**Review by xylar — COMMENTED** [review]
Wow, this is amazing!  And nearly there.  I just have a few small pieces of clean-up I'd recommend.  Then, this will need to be rebased because of the TMIX PR that just went in (perhaps among other conflicts).  And then it's ready to go to E3SM!

Happy to help with the rebase if you need.

File: components/mpas-ocean/src/shared/mpas_ocn_forcing.F (hunk 7 above)
Comment (irenavankova) [reply components/mpas-ocean/src/shared/mpas_ocn_forcing.F:187]: What is an appropriate way to do this? to raise error and kill the run?

File: components/mpas-ocean/src/tracer_groups/Registry_ecosys.xml (hunk 9 above)
Comment (irenavankova) [reply components/mpas-ocean/src/tracer_groups/Registry_ecosys.xml:413]: I told him about it a while ago, but I think he forgot about it

File: components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F (hunk 8 above)
Comment (irenavankova) [reply components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:3509]: Done

File: components/mpas-ocean/src/shared/mpas_ocn_forcing.F (hunk 7 above)
Comment (irenavankova) [reply components/mpas-ocean/src/shared/mpas_ocn_forcing.F:187]: Done

File: components/mpas-ocean/src/Registry.xml (hunk 6 above)
Comment (irenavankova) [reply components/mpas-ocean/src/Registry.xml:3742]: Done

File: components/mpas-ocean/bld/namelist_files/namelist_definition_mpaso.xml (hunk 5 above)
Comment (irenavankova) [reply components/mpas-ocean/bld/namelist_files/namelist_definition_mpaso.xml]: Done

File: components/mpas-ocean/bld/build-namelist-section (hunk 4 above)
Comment (irenavankova) [reply components/mpas-ocean/bld/build-namelist-section]: done

**Comment by irenavankova** [comment]
I rebased to master and moved to E3SM Project as new branched named ocn/add-subglacial-runoff

**Comment by irenavankova** [comment]
I rebased to master and moved to E3SM Project as new branched named ocn/add-subglacial-runoff
